[WIP] 17/fx data

sgratiy opened the pull request.

NileGraddis commented on src/main/python/dictfield.py:
```diff
@@ -0,0 +1,25 @@
+from marshmallow.fields import Field
+
+
+class DictField(Field):
```
Why not `Nested`?

sgratiy replied:
Because the format of json is such that keys are sweep numbers (that change)
    "sweep_features": {
      "11": {
        "spikes": [
          {

NileGraddis replied:
https://marshmallow.readthedocs.io/en/stable/api_reference.html#marshmallow.fields.Dict?

sgratiy replied:
Actually Dict works, thanks.

NileGraddis commented on src/main/python/fx_data.py:
```diff
@@ -0,0 +1,47 @@
+
+        self.input_nwb_file = nwb_path
+        self.stimulus_ontology_file = ontology_file
+        self.output_nwb_file = os.path.join(str(Path.cwd()), "output.nwb")
```
per in-person, we should avoid writing out data unless we have a reason to do so. If we must, we should use a temporary directory rather than cwd.

NileGraddis commented on src/main/python/fx_data.py:
```diff
@@ -0,0 +1,47 @@
+        self._state_out_of_date = True
+
+        self.input_nwb_file = nwb_path
+        self.stimulus_ontology_file = ontology_file
```
any reason to use the ontology_file vs the ontology itself?

NileGraddis commented on src/main/python/pre_fx_data.py:
```diff
@@ -12,11 +12,11 @@
 from ipfx.stimulus import StimulusOntology, Stimulus
 from ipfx.data_set_utils import create_data_set
-
+import ipfx.sweep_props as sp
```
Can you import the items you need from sweep props? The two-letter module imports can be annoying to read / search unless they are very well known (e.g. np).

NileGraddis commented on src/main/python/pre_fx_data.py:
```diff
@@ -292,6 +295,36 @@ def run_extraction_and_auto_qc(self, nwb_path, stimulus_ontology, qc_criteria, c
+
+    def get_passing_sweep_features(self):
+        manual_sweep_states = self.get_non_default_manual_sweep_states()
+        print(manual_sweep_states)
```
intentional?

NileGraddis commented on src/main/python/schemas.py:
```diff
@@ -27,13 +28,234 @@ class QcCriteria(Schema):
 
 
-class PipelineInput(Schema):
+class SweepParameters(Schema):
```
nice!

sgratiy committed:
Add schema for the feature extraction output

sgratiy committed:
PreFxData emits signal on manual change in sweeep state

Update sweep state on manual qc change
Emit signal with parameters needed by the feature extractor

sgratiy committed:
Add FxData

FxData gets out of date when manual qc state is changed
FxData can compute features

NileGraddis reviewed (approved):
lgtm :shark:
